Wrap track extender/initializer functions inside concrete Actions

drbenmorgan opened the pull request:
Per #662, this is an initial go at wrapping the track initializer/extender functions from `TrackInitUtils.hh` in concrete classes of `ExplicitActionInterface`. As discussed in the Issue, actions take a `const` reference to `CoreRef<M>`, but the track functions require a non-const reference for some of the operations (mainly resizing track state collections). The PR as submitted is then split into two parts:

- Update track initializer function interfaces to take `CoreRef<M> const&`.
  - Use `const_cast` on the data when a non-const reference is required. Kept as local as possible so it's clear where this is needed for later refactoring/etc.
- Wrapping of `initialize_tracks` and `extend_from_secondaries` functions inside concrete instances of `ExplicitActionInterface`.
  - The `extend_from_primaries` function is not wrapped yet as this takes both a `CoreRef<M>` and a collection of `Primary`ies, so needs further thought.

No tests, updates to other areas yet just to check that this basic structure/implementation is o.k. for what's needed by #662. I'd guess tests would mean providing the `TrackInit` tests implemented in terms of the Actions?

sethrj commented:
CI test failure is that weird intermittent Geant4 on HIP...

drbenmorgan commented:
O.k., a very basic test added, though this isn't really anymore than a slightly modded version of the `TestTrackInitSecondaryTest, secondaries` fixture. Let me know if anything else is needed in this fixture, or adding extra tests.

Everything seems to pass on a local Linux/CUDA build, but let's see what happens on HIP again!

sethrj commented on src/celeritas/track/InitializeTracksAction.hh:
```diff
@@ -0,0 +1,74 @@
+//---------------------------------------------------------------------------//
+
+inline void
+InitializeTracksAction::execute([[maybe_unused]] CoreDeviceRef const& core) const
```
I have a mild preference for putting these into a small `.cc` file to hide the `TrackInitUtils` include.

sethrj reviewed (commented):
Sorry for the delay in reviewing (as I said, busy week with ATLAS). This looks good so far; but as part of the MR can you also integrate it into the stepping loop? I think you'll just need to add the action inside the `CoreParams::CoreParams` constructor (like we do with the geometry actions) and remove the explicit call in `Stepper<M>::operator()`.

sethrj commented:
> The only issue here is that `Stepper<M>::operator()` stores and returns counts of track statuses:
> 
> `alive` and `queued` are o.k., but the `active` value is recorded immediately after the track initialization action. I'm not sure of the best solution here, though clearly the value of `core_ref_.states.init.vacancies.size()` needs storing at the end of/after `InitializeTracksAction::execute` somewhere. Could be a data member of `InitializeTracksAction`, but then there's a coupling to the action...

I think the `init.vacancies.size` will only be affected by `initialize_tracks` and `extend_from_secondaries`. So until we create an action from `extend_from_secondaries`, I think this will still work?

drbenmorgan commented:
As a first step to Action-ize things, I've just added the `ExtendFromSecondaries` action to `CoreParams` as the alive/queued values are always calculated after completion of the sequence execution. 

Unless I'm missing something fundamental, I don't think Action-izing the track initialization step and calculating "active" can be done directly due to the lines:

https://github.com/celeritas-project/celeritas/blob/277c8f7d1807bb195470e8218b00796fac6467ab/src/celeritas/global/Stepper.cc#L74-L77

The calculation of `result.active` would either need an extra action to do the count after the action runs, or, we'd have to do the calculation before execution of the action sequence using the same logic as in `initialize_tracks`?

sethrj commented:
@drbenmorgan Quite so. I think the most sensible thing would be to add a `size_type active;` that gets updated by the `InitializePrimaries` action. That way we could just copy out that value at the end of the stepping loop.

This might also need to change for some work that @esseivaju @amandalund and I are about to get started on, which is changing from the easy/lazy/inefficient strategy of "masking" inactive threads to remapping "track slots" to "threads": https://github.com/celeritas-project/celeritas/issues/675

drbenmorgan commented:
Thanks @sethrj, I _think_ I've implemented this per your suggestion, though done via a data member in `TrackInitStateData` and updated by `initialise_tracks` (very similar to the secondary counter).

Noted on the #675 work - fine on my side to wait for that to complete/merge before proceeding here if that's easier.

sethrj commented:
Thanks @drbenmorgan . I think we just need to resolve the conflicts in the updated TrackInit test (I can help with this if you want) and then we'll call this good to go! I'd like to get this in before @esseivaju starts the next step of the slot refactor.

drbenmorgan commented:
Sounds good - are you o.k. with a rebase onto `master` at this point?

drbenmorgan commented:
Oh hold on, GitHub lets me do this interactively with a merge, so all good I think, just going to test locally as well.

drbenmorgan commented:
Urghh, messed that up, but conflict should now be resolved _correctly_, and build and tests pass locally.

sethrj reviewed (approved):
Nice work Ben, and thanks! It'll be good to have this extra timing granularity (up till now our "action fraction" pie charts have missed this).
